# MetaG -the metagenomics pipeline
## What is MetaG?
MetaG is a pipeline for the analysis of marker genes (16S, 28S rRNA) and for the
analysis of whole genomes of viruses. It can handle the distinct sequencing
error profiles of short-read and long-read sequencing data.
MetaG has been shown to perform better than state-of-the-art classifiers,
especially at lower taxonomic ranks.

Apart from its outstanding performance, MetaG was designed for easy usage.
This is realized by an installation/configuration via an interactive installer, as
well as a thorough documentation which is also available in man page format.

If you want to learn more about MetaG, read its preprint: doi.org/foo/bar (coming soon)
The program was tested on Ubuntu 18.04.3 LTS, macOS 10.15.3 and FreeBSD 12.1.

Free yourself from hardware restrictions: Use MetaG online!
http://www.bioinformatics.uni-muenster.de/tools/metag

## Directory and files
### metag_src
Download this folder to install MetaG locally. See README in this directory
to learn how to get started.

### supplemental
Supplemental files and scripts for the publication. See README in this directory
for further details.
